integration/spring-boot/autoconfigure/src/main/java/org/apache/cxf/spring/boot/autoconfigure/micrometer/MicrometerMetricsAutoConfiguration.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.cxf.metrics.micrometer.MicrometerMetricsProvider;
+import org.apache.cxf.metrics.micrometer.provider.TagsProvider;
+import org.apache.cxf.metrics.micrometer.provider.TimedAnnotationProvider;
+import org.apache.cxf.metrics.micrometer.provider.DefaultExceptionClassProvider;
+import org.apache.cxf.metrics.micrometer.provider.ExceptionClassProvider;
+import org.apache.cxf.metrics.micrometer.provider.jaxws.DefaultJaxwsFaultCodeProvider;
+import org.apache.cxf.metrics.micrometer.provider.jaxws.JaxwsFaultCodeProvider;
+import org.apache.cxf.metrics.micrometer.provider.jaxws.JaxwsTags;
+import org.apache.cxf.metrics.micrometer.provider.jaxws.JaxwsTagsProvider;
+import org.apache.cxf.spring.boot.autoconfigure.CxfProperties;
+import org.apache.cxf.spring.boot.autoconfigure.CxfProperties.Metrics.Server;
+import org.apache.cxf.spring.boot.autoconfigure.micrometer.provider.jaxws.SpringBasedTimedAnnotationProvider;
+import org.springframework.boot.actuate.autoconfigure.metrics.MetricsAutoConfiguration;
+import org.springframework.boot.actuate.autoconfigure.metrics.OnlyOnceLoggingDenyMeterFilter;
+import org.springframework.boot.actuate.autoconfigure.metrics.export.simple.SimpleMetricsExportAutoConfiguration;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.annotation.Order;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.config.MeterFilter;
+
+@Configuration
+@AutoConfigureAfter({MetricsAutoConfiguration.class, SimpleMetricsExportAutoConfiguration.class})
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
+@ConditionalOnClass({JaxWsServerFactoryBean.class, MetricsProvider.class})
+@ConditionalOnBean({MeterRegistry.class})
+@EnableConfigurationProperties(CxfProperties.class)
+public class MicrometerMetricsAutoConfiguration {
+
+    private final CxfProperties properties;
+
+    public MicrometerMetricsAutoConfiguration(CxfProperties properties) {
+        this.properties = properties;
+    }
+
+    @Bean
+    public TimedAnnotationProvider timedAnnotationProvider() {
+        return new SpringBasedTimedAnnotationProvider();
+    }
+
+    @Bean
+    public JaxwsTags jaxwsTags() {
+        return new JaxwsTags();
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(ExceptionClassProvider.class)
+    public ExceptionClassProvider exceptionClassProvider() {
```

Comment:
I would suggest to remove `ExceptionClassProvider`  and `JaxwsFaultCodeProvider` altogether and replace it with something like `TagsCustomizer`. 
```
interface TagsCustomizer() {
    Iterable<Tag> getAdditionalTags(Exchange ex);
}
```

There could many of them (fe, could be one for JAX-WS and one for JAX-RS) and `MicrometerMetricsContext` could apply the available customizers before emitting the final set of tags. 

```
Supplier<Iterable<Tag>> tags = () -> this.tagsProvider.getTags(ex) + <all tags from customizers>
```

What do you think?

Reply:
Fair point, it leaves open some possibilities.